hpl-2.2/man/man3/HPL_grid_info.3
.TH HPL_grid_info 3 "February 24, 2016" "HPL 2.2" "HPL Library Functions"
.SH NAME
HPL_grid_info \- Retrieve grid information.
.SH SYNOPSIS
\fB\&#include "hpl.h"\fR
 
\fB\&int\fR
\fB\&HPL_grid_info(\fR
\fB\&const HPL_T_grid *\fR
\fI\&GRID\fR,
\fB\&int *\fR
\fI\&NPROW\fR,
\fB\&int *\fR
\fI\&NPCOL\fR,
\fB\&int *\fR
\fI\&MYROW\fR,
\fB\&int *\fR
\fI\&MYCOL\fR
\fB\&);\fR
.SH DESCRIPTION
\fB\&HPL_grid_info\fR
returns  the grid shape and the coordinates in the grid
of the calling process.  Successful  completion  is  indicated by the
returned error code  MPI_SUCCESS. Other error codes depend on the MPI
implementation.
.SH ARGUMENTS
.TP 8
GRID    (local input)           const HPL_T_grid *
On entry,  GRID  points  to the data structure containing the
process grid information.
.TP 8
NPROW   (global output)         int *
On exit,   NPROW  specifies the number of process rows in the
grid. NPROW is at least one.
.TP 8
NPCOL   (global output)         int *
On exit,   NPCOL  specifies  the number of process columns in
the grid. NPCOL is at least one.
.TP 8
MYROW   (global output)         int *
On exit,  MYROW  specifies my  row process  coordinate in the
grid. MYROW is greater than or equal  to zero  and  less than
NPROW.
.TP 8
MYCOL   (global output)         int *
On exit,  MYCOL specifies my column process coordinate in the
grid. MYCOL is greater than or equal  to zero  and  less than
NPCOL.
.SH SEE ALSO
.BR HPL_pnum \ (3),
.BR HPL_grid_init \ (3),
.BR HPL_grid_exit \ (3).


hpl-2.2/HISTORY
==============================================================
 High Performance Computing Linpack Benchmark (HPL)
 HPL - 2.2 - February 24, 2016
==============================================================

 History

 - 09/09/00 Public release of Version 1.0

 - 09/27/00 A couple of mistakes in the  VSIPL  port have been
 corrected.  The tar file as well as the web site were updated
 on September 27th, 2000.  Note  that  these problems were not
 affecting the BLAS version of the software in any way.

 - 01/01/04 Version 1.0a
 The  MPI  process grid  numbering  scheme  is now an run-time
 option.
 The inlined assembly  timer  routine that caused the compila-
 tion to fail when using  gcc  version 3.3  and above has been
 removed from the package.
 Various building problems on the T3E have been fixed;  Thanks
 to Edward Anderson.

 - 15/12/04 Version 1.0b
 Weakness of the pseudo-random matrix generator found for pro-
 blem sizes being power of twos and larger  than 2^15;  Thanks
 to Gregory Bauer. This problem has not been fixed. It is thus
 currently recommended to  HPL  users willing to test matrices
 of size larger than 2^15 to not use power twos.

 When the matrix size is such that one needs  > 16 GB  per MPI
 rank,  the  intermediate  calculation  (mat.ld+1) * mat.nq in
 HPL_pdtest.c  ends up  overflowing  because  it is done using
 32-bit arithmetic.  This issue has been fixed by  typecasting
 to size_t; Thanks to John Baron.

 - 09/10/08 Version 2.0

 Piotr Luszczek changed to 64-bit RNG, modified files:
 -- [M] include/hpl_matgen.h
 -- [M] testing/matgen/HPL_ladd.c
 -- [M] testing/matgen/HPL_lmul.c
 -- [M] testing/matgen/HPL_rand.c
 -- [M] testing/ptest/HPL_pdinfo.c

 For a motivation for the change, see:
    Dongarra and Langou, ``The Problem with the Linpack
    Benchmark Matrix Generator'', LAWN 206, June 2008.

 -- [M] testing/ptest/HPL_pdtest.c  --

 Julien Langou changed the test for correctness from
      ||Ax-b||_oo / ( eps * ||A||_1  * N            )
      ||Ax-b||_oo / ( eps * ||A||_1  * ||x||_1      )
      ||Ax-b||_oo / ( eps * ||A||_oo * ||x||_oo * N )
 to the normwise backward error
      || r ||_oo / ( eps * ( || x ||_oo * || A ||_oo + || b ||_oo ) * N )
 See:
  Nicholas J. Higham, ``Accuracy and Stability of Numerical Algorithms'',
  Society for Industrial and Applied Mathematics, Philadelphia, PA, USA,
  Second Edition, pages = xxx+680, ISBN = 0-89871-521-0, 2002.

 Note that in our case || b ||_oo is almost for sure
 1/2, we compute it anyway.

 - 10/26/2012 Version 2.1

 Piotr Luszczek introduced exact time stamping for HPL_pdgesv():
 -- [M] dist/include/hpl_misc.h
 -- [M] dist/testing/ptest/HPL_pdtest.c

 Piotr Luszczek fixed out-of-bounds access in data spreading functions
 and exact time stamping for HPL_pdgesv():
 -- [M] dist/src/pgesv/HPL_spreadN.c
 -- [M] dist/src/pgesv/HPL_spreadT.c
 Thanks to Stephen Whalen from Cray.

 - 02/24/2016 Version 2.2

 Piotr Luszczek added continuous reporting of factorization progress
 submitted by Intel and make scripts that uses Intel software tools and
 libraries and their Apple's Mac OS X equivalents.


hpl-2.2/src/grid/HPL_grid_info.c
/* 
 * -- High Performance Computing Linpack Benchmark (HPL)                
 *    HPL - 2.2 - February 24, 2016                          
 *    Antoine P. Petitet                                                
 *    University of Tennessee, Knoxville                                
 *    Innovative Computing Laboratory                                 
 *    (C) Copyright 2000-2008 All Rights Reserved                       
 *                                                                      
 * -- Copyright notice and Licensing terms:                             
 *                                                                      
 * Redistribution  and  use in  source and binary forms, with or without
 * modification, are  permitted provided  that the following  conditions
 * are met:                                                             
 *                                                                      
 * 1. Redistributions  of  source  code  must retain the above copyright
 * notice, this list of conditions and the following disclaimer.        
 *                                                                      
 * 2. Redistributions in binary form must reproduce  the above copyright
 * notice, this list of conditions,  and the following disclaimer in the
 * documentation and/or other materials provided with the distribution. 
 *                                                                      
 * 3. All  advertising  materials  mentioning  features  or  use of this
 * software must display the following acknowledgement:                 
 * This  product  includes  software  developed  at  the  University  of
 * Tennessee, Knoxville, Innovative Computing Laboratory.             
 *                                                                      
 * 4. The name of the  University,  the name of the  Laboratory,  or the
 * names  of  its  contributors  may  not  be used to endorse or promote
 * products  derived   from   this  software  without  specific  written
 * permission.                                                          
 *                                                                      
 * -- Disclaimer:                                                       
 *                                                                      
 * THIS  SOFTWARE  IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,  INCLUDING,  BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE UNIVERSITY
 * OR  CONTRIBUTORS  BE  LIABLE FOR ANY  DIRECT,  INDIRECT,  INCIDENTAL,
 * SPECIAL,  EXEMPLARY,  OR  CONSEQUENTIAL DAMAGES  (INCLUDING,  BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA OR PROFITS; OR BUSINESS INTERRUPTION)  HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT,  STRICT LIABILITY,  OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
 * ---------------------------------------------------------------------
 */ 
/*
 * Include files
 */
#include "hpl.h"

#ifdef STDC_HEADERS
int HPL_grid_info
(
   const HPL_T_grid *               GRID,
   int *                            NPROW,
   int *                            NPCOL,
   int *                            MYROW,
   int *                            MYCOL
)
#else
int HPL_grid_info
( GRID, NPROW, NPCOL, MYROW, MYCOL )
   const HPL_T_grid *               GRID;
   int *                            NPROW;
   int *                            NPCOL;
   int *                            MYROW;
   int *                            MYCOL;
#endif
{
/* 
 * Purpose
 * =======
 *
 * HPL_grid_info returns  the grid shape and the coordinates in the grid
 * of the calling process.  Successful  completion  is  indicated by the
 * returned error code  MPI_SUCCESS. Other error codes depend on the MPI
 * implementation.
 *
 * Arguments
 * =========
 *
 * GRID    (local input)                 const HPL_T_grid *
 *         On entry,  GRID  points  to the data structure containing the
 *         process grid information.
 *
 * NPROW   (global output)               int *
 *         On exit,   NPROW  specifies the number of process rows in the
 *         grid. NPROW is at least one.
 *
 * NPCOL   (global output)               int *
 *         On exit,   NPCOL  specifies  the number of process columns in
 *         the grid. NPCOL is at least one.
 *
 * MYROW   (global output)               int *
 *         On exit,  MYROW  specifies my  row process  coordinate in the
 *         grid. MYROW is greater than or equal  to zero  and  less than
 *         NPROW.
 *
 * MYCOL   (global output)               int *
 *         On exit,  MYCOL specifies my column process coordinate in the
 *         grid. MYCOL is greater than or equal  to zero  and  less than
 *         NPCOL.
 *
 * ---------------------------------------------------------------------
 */ 
/*
 * .. Local Variables ..
 */
   *NPROW = GRID->nprow; *NPCOL = GRID->npcol;
   *MYROW = GRID->myrow; *MYCOL = GRID->mycol;
   return( MPI_SUCCESS );
/*
 * End of HPL_grid_info
 */
}
